dask/array/slicing.py

```diff
@@ -566,6 +567,12 @@ def take(outname, inname, chunks, index, axis=0):
     if not np.isnan(chunks[axis]).any():
         from dask.array._shuffle import _shuffle
 
+        try:
+            # Handle cupy case
+            index = to_numpy_dispatch(index)
```

Comment:
>This pattern seems pretty annoying

Yes, I agree!

The underlying problem is that cupy will raise an error when you do something that "implicitly" requires device-to-host transfer. The idea is that you should explicity move the data from device to host if that's what you really want.

I think the use of `to_numpy_dispatch` is a fairly simple solution to this. However, I put it in a try/except block for now, since we haven't registered a dispatch for anything other than `cupy.ndarray` and `numpy.ndarray`. That said, I don't think we want to copy this pattern everywhere (because I agree that it's "annoying").

What do you think would be the "least annoying" pattern here? Maybe we simply define the necessary dispatch logic in 'backends.py' so this can be one line? Maybe we use a single `as_numpy_array` utility everywhere that we want to explicitly cast to numpy from other possible backends?

Reply:
I registered simple logic for `list` and `tuple` so this should only be one like now.

Just a note that not being able to use `np.asarray` explicit isn't a new consideration in `dask.array`. There are various `asarray` utilities for this in `dask.array.utils`.

What makes this particular case "special" is that I am not trying to "fix" the breaking changes made recently to `take`. Rather, I'm suggesting that we simply cast everything to numpy for this particular function for now. After gpuCI is back up and passing, I can try to add cupy support (if necessary).

Reply:
I feel `np.asarray` feels like a pretty _explicit_ thing and I'm a little surprised about cupy behavior here.

`to_numpy_dispatch` isn't used anywhere in the code base from what I can tell

Reply:
>I feel np.asarray feels like a pretty explicit thing and I'm a little surprised about cupy behavior here.

Yeah, I can understand that, but the behavior is certainly not new. Numpy uses the array protocol to dispatch compute behavior. That makes it safe to use functions like `np.mean` with cupy arrays. I believe the idea is to make sure users don't *silently* move their data with common operations like `np.asarray`.

>to_numpy_dispatch isn't used anywhere in the code base from what I can tell

You are correct. We typically try to **keep** cupy arrays in device memory. This PR is just a short-term fix to get gpuCI back up since recent `dask.array` PRs made this a bit tricky. The proper fix may not be that hard, I just haven't had much time to try yet.

Reply:
Sorry - I was conflating the failing gpuCI tests with other (uncovered) cupy problems. The gpuCI tests should be resolved by adding a few missing `like=` arguments to keep cupy data on device. My mistake (and thanks to @pentschev for reminding me offline)